vm/devices/support/fs/lxutil/src/windows/util.rs

```diff
@@ -886,3 +1258,106 @@ fn set_time_to_timespec(time: &SetTime) -> lx::Timespec {
+    mode: &mut lx::mode_t,
+    owner_gid: &mut lx::gid_t,
+) {
+    if parent_mode & lx::S_ISGID != 0 {
+        if lx::s_isdir(*mode) {
+            *mode |= lx::S_ISGID;
+        }
+
+        *owner_gid = parent_gid;
+    }
+}
+
+/// Convert a reparse tag into the appropriate file mode
+pub fn reparse_tag_to_file_mode(reparse_tag: u32) -> lx::mode_t {
+    // We need to redefine some of the tags, in the windows crate they're defined as i32s
+    // instead of u32s. In the headers themselves they're just #define directives
+    const IO_REPARSE_TAG_LX_SYMLINK: u32 = FileSystem::IO_REPARSE_TAG_LX_SYMLINK as u32;
+    const IO_REPARSE_TAG_LX_FIFO: u32 = FileSystem::IO_REPARSE_TAG_LX_FIFO as u32;
+    const IO_REPARSE_TAG_LX_CHR: u32 = FileSystem::IO_REPARSE_TAG_LX_CHR as u32;
+    const IO_REPARSE_TAG_LX_BLK: u32 = FileSystem::IO_REPARSE_TAG_LX_BLK as u32;
+
+    match reparse_tag {
+        W32Ss::IO_REPARSE_TAG_SYMLINK
+        | W32Ss::IO_REPARSE_TAG_MOUNT_POINT
+        | IO_REPARSE_TAG_LX_SYMLINK => lx::S_IFLNK,
+        W32Ss::IO_REPARSE_TAG_AF_UNIX => lx::S_IFSOCK,
+        IO_REPARSE_TAG_LX_FIFO => lx::S_IFIFO,
+        IO_REPARSE_TAG_LX_CHR => lx::S_IFCHR,
+        IO_REPARSE_TAG_LX_BLK => lx::S_IFBLK,
+        _ => 0,
+    }
+}
+
+/// Convert a file mode to the appropriate reparse tag
+pub fn file_mode_to_reparse_tag(file_mode: lx::mode_t) -> u32 {
+    match file_mode & lx::S_IFMT {
+        lx::S_IFIFO => FileSystem::IO_REPARSE_TAG_LX_FIFO as _,
+        lx::S_IFSOCK => W32Ss::IO_REPARSE_TAG_AF_UNIX,
+        lx::S_IFCHR => FileSystem::IO_REPARSE_TAG_LX_CHR as _,
+        lx::S_IFBLK => FileSystem::IO_REPARSE_TAG_LX_BLK as _,
+        lx::S_IFLNK => FileSystem::IO_REPARSE_TAG_LX_SYMLINK as _,
+        _ => W32Ss::IO_REPARSE_TAG_RESERVED_ZERO as _,
+    }
+}
+
+/// Convert a reparse tag to the appropriate DT_* file type.
+pub fn reparse_tag_to_file_type(reparse_tag: u32) -> u8 {
+    // We need to redefine some of the tags, in the windows crate they're defined as i32s
+    // instead of u32s. In the headers themselves they're just #define directives
+    const IO_REPARSE_TAG_LX_SYMLINK: u32 = FileSystem::IO_REPARSE_TAG_LX_SYMLINK as _;
+    const IO_REPARSE_TAG_LX_FIFO: u32 = FileSystem::IO_REPARSE_TAG_LX_FIFO as _;
+    const IO_REPARSE_TAG_LX_CHR: u32 = FileSystem::IO_REPARSE_TAG_LX_CHR as _;
+    const IO_REPARSE_TAG_LX_BLK: u32 = FileSystem::IO_REPARSE_TAG_LX_BLK as _;
+
+    match reparse_tag {
+        W32Ss::IO_REPARSE_TAG_SYMLINK
+        | W32Ss::IO_REPARSE_TAG_MOUNT_POINT
+        | IO_REPARSE_TAG_LX_SYMLINK => lx::DT_LNK,
+        W32Ss::IO_REPARSE_TAG_AF_UNIX => lx::DT_SOCK,
+        IO_REPARSE_TAG_LX_FIFO => lx::DT_FIFO,
+        IO_REPARSE_TAG_LX_CHR => lx::DT_CHR,
+        IO_REPARSE_TAG_LX_BLK => lx::DT_BLK,
+        _ => lx::DT_UNK,
+    }
+}
+
+/// Check if a UnicodeString is "." or ".."
+pub fn is_self_relative_unicode_path(path: &windows::UnicodeString) -> bool {
+    let slice = path.as_slice();
```

Comment:
Can this whole thing be `const DOT: u16 = '.' as u16; matches!(path.as_slice(), [DOT] | [DOT, DOT])` ?